src/com/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -883,6 +884,106 @@
     (sql/do-commands
      "DROP INDEX fact_values_string_trgm")))
 
+(defn- lift-fact-paths-into-facts
+  "Pairs paths and values directly in facts, i.e. change facts from (id
+  value) to (id path value)."
+  []
+  (sql/do-commands
+   (if (scf-utils/postgres?)
+     "SET CONSTRAINTS ALL DEFERRED" ;; In case it allows additional optimization?
+     "SELECT 1")
+
+   ;; Build complete facts table as of migration 28.
+   "CREATE TABLE facts_transform
+        (factset_id bigint NOT NULL,
+         fact_path_id bigint NOT NULL,
+         fact_value_id bigint NOT NULL)"
+
+   ;; Needed for the insert if nothing else.
+   "CREATE INDEX fact_paths_path_idx on fact_paths(path)"
+   "CREATE INDEX fact_values_value_hash_idx on fact_values(value_hash)"
+
+   ;; Patch up facts refrences to refer to the min id path/value
+   ;; wherever there's more than one option.
+   "INSERT INTO facts_transform (factset_id, fact_path_id, fact_value_id)
```

Comment:
To tie up the loose ends in the above comment, I had a way large fact_path table. 430k fact paths is pretty ridiculous. Bumping that down to a reasonable (but still high) number of 500 (with 100k fact_values), I do get the migration to finish. It took about 2 minutes on my laptop.

I thought this was just sensitive to the size of the fact_paths table, but this looks to be sensitive to the size of the fact_values table as well. I increased the fact values to 195k (fact_paths still ~500) and migrated. I eventually killed it after 20 minutes. It looks like it's sensitive to larger fact_values as well. Although 195k fact_values is high, seems likely we have users with that amount (or more). My guess is, it's the select mins from the facts_transform population query that takes so long.

I think I've [found another way](https://github.com/senior/puppetdb/commit/8cda94e6dbf8ab105b4ddbacef8948e1727fd932). I created two tables, one that has the factset_id, path and value_hash. With that populated, I deleted the path/value duplicates (same code as before, just moved up in the migration). From this, the data can be moved from this intermediate table to the facts_transform table with the correct PKs using a simple join, no aggregates. We can drop the facts table once the first intermediate table is loaded, and drop the first intermediate after the second is loaded. There's only a small increase in size from what was there previously (the value_hash and path mainly).

I ran the migration on the same dataset as the one I killed after 20 minutes and it ran in just under 10 seconds.